compiler/rustc_expand/src/mbe/metavar_expr.rs

```diff
@@ -41,6 +45,16 @@ impl MetaVarExpr {
         check_trailing_token(&mut tts, sess)?;
         let mut iter = args.trees();
         let rslt = match ident.as_str() {
+            "concat" => {
+                let lhs_sign = try_eat_dollar(&mut iter);
+                let lhs = parse_ident(&mut iter, sess, ident.span)?;
+                if !try_eat_comma(&mut iter) {
+                    return Err(sess.span_diagnostic.struct_span_err(ident.span, "expected comma"));
+                }
+                let rhs_sign = try_eat_dollar(&mut iter);
+                let rhs = parse_ident(&mut iter, sess, ident.span)?;
+                MetaVarExpr::Concat((lhs, lhs_sign), (rhs, rhs_sign))
+            }
```

Comment:
Does this distinguish between `${concat($tt_that_is_actually_dollar ident, ...)}` and `${concat{$ident, ...}}`?

Reply:
Yes, you can see such scenario in the newly added tests.